KosherCocoa/Library/Core/Solar/KCNOAACalculator.m

```diff
@@ -0,0 +1,450 @@
+//
+//  KCNOAACalculator.m
+//  KosherCocoa
+//
+//  Created by Elyahu on 1/23/23.
+//
+
+#import "KCNOAACalculator.h"
+#import "trigonometry.h"
+
+@implementation KCNOAACalculator
+
+- (id)initWithGeoLocation:(KCGeoLocation *)geolocation
+{
+    
+    self = [super init];
+    
+    if (self)
+    {
+        [self setCalculatorName:@"US National Oceanic and Atmospheric Administration Algorithm"];
+        
+        //
+        //  If there's no GeoLocation, then
+        //  create one for where the equator
+        //  and the prime meridian meet.
+        //
+        
+        
+        if (!geolocation)
+        {
+            
+            //
+            //  Create the default KCGeoLocation
+            //
+            
+            KCGeoLocation *tempG = [[KCGeoLocation alloc] initWithName:@"Default" andLatitude:0.0 andLongitude:0.0 andTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
+            
+            //
+            //  Assign it to the geoLoc object
+            //
+            
+            geolocation = tempG;
+            
+        }
+        
+        //
+        //  Store the geoLocation object
+        //
+        
+        _geoLocation = geolocation;
+    }
+    
+    return self;
+}
+
+
+
+/**
+ *  A method returning the sunrise or sunset in UTC as a double. If an error was
+ *  encountered in the calculation (as expected in some locations, such
+ *  as near the poles) NAN will be returned.
+ *
+ *  @param year A Gregorian year number as an integer.
+ *
+ *  @param month The month of the gregorian year. Valud values range from 1 to 12.
+ *
+ *  @param day The day of the gregorian month. Valid values range from 1 to 31.
+ *
+ *  @param longitude The longitude in degrees. (Longitude values west of Meridian are negative.)
+ *
+ *  @param latitude The latitude in degrees. longitudes south of Equator are negative
+ *
+ *  @param zenith Which zenith to use in the calculation.
+ *
+ *  @param type The type of calculation to carry out. Values are kCalcTypeSunrise or kCalcTypeSunset. (Alternatively, an integer 0 or 1)
+ *
+ *  @return The time that sunset occurs UTC as a double.
+ */
+
+
+- (double)sunriseOrSunsetForYear:(NSInteger)year andMonth:(NSInteger)month andDay:(NSInteger)day atLongitude:(double)longitude andLatitude:(double)latitude withZenith:(double)zenith andType:(NSInteger)type
```

Comment:
Is this method un-implemented in the other calendar or did we not implement it here? This is implemented in [`KCSunriseAndSunsetCalculator.m`][1].

[1]: https://github.com/MosheBerman/KosherCocoa/blob/dea2e03d0d60f3334dda16d20e5d4b92f972366f/KosherCocoa/Library/Core/Solar/KCSunriseAndSunsetCalculator.m]

Reply:
So, this method is a bit tricky for me. It was not in the Original Java file for the NOAACalculator. In order to get a head start, I copied the code (both .h and .m files) from the KCSunriseAndSunsetCalculator, and I edited them. So this method may not be needed at all, but I kept it in because the Class needed to implement it. I would just delete the method if it is not going to be used, or if this method was just needed for the KCSunriseAndSunsetCalculator.

Reply:
If it's not called anywhere, we can remove it.